src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -429,6 +429,417 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
+One round of the division algorithm, that tries to perform a subtract shift.
+Note that this is only called when `r.msb = false`, so we will not overflow.
+-/
+def divSubtractShift (n : BitVec w) (d : BitVec w) (wn : Nat) (qr : DivModState w) :
+    DivModState w :=
+  let r' := shiftConcat qr.r (n.getLsb (wn - 1)) -- if r ≥ d, then we have a quotient bit.
+  if r' < d
+  then {
+    q := qr.q.shiftConcat false, -- If `r' < d`, then we do not have a quotient bit.
+    r := r'
+  } else {
+    q := qr.q.shiftConcat true, -- If `r' ≥ d`, then we have a quotient bit.
+    r := r' - d -- we subtract to maintain the invariant that `r < d`.
+  }
+
+/-- The value of shifting by `wn - 1` equals shifting by `wn` and grabbing the lsb at `(wn - 1)`. -/
+theorem DivModState.toNat_shiftRight_sub_one_eq
+    (qr : DivModState w) (h : qr.LawfulShiftSubtract wr wn n d):
+    n.toNat >>> (wn - 1) = (n.toNat >>> wn) * 2 + (n.getLsb (wn - 1)).toNat := by
+  have hn := shiftRight_sub_one_eq_shiftConcat_getLsb_of_lt (n := n) (wn := wn) h.hwn_lt
+  obtain hn : (n >>> (wn - 1)).toNat = ((n >>> wn).shiftConcat (n.getLsb (wn - 1))).toNat := by
+    simp [hn]
+  simp only [toNat_ushiftRight] at hn
+  rw [toNat_shiftConcat_eq_of_lt_of_lt_two_pow (k := w - wn)] at hn
+  · rw [hn]
+    rw [toNat_ushiftRight]
+  · have := h.hwn_lt
+    have := h.hw
+    omega
+  · apply BitVec.toNat_ushiftRight_lt
+    have := h.hwrn
+    omega
+
+/--
+This is used when proving the correctness of the divison algorithm,
+where we know that `r < d`.
+We then want to show that `((r.shiftConcat b) - d) < d` as the loop invariant.
+In arithmethic, this is the same as showing that
+`r * 2 + 1 - d < d`, which this theorem establishes.
+-/
+private theorem two_mul_add_sub_lt_of_lt_of_lt_two (h : a < x) (hy : y < 2) :
+    2 * a + y - x < x := by omega
+
+/-- We show that the output of `divSubtractShift` is lawful, which tells us that it
+obeys the division equation. -/
+def divSubtractShiftProof (qr : DivModState w) (h : qr.LawfulShiftSubtract  wr wn n d) :
+    DivModState.Lawful w (wr + 1) (wn - 1) n d (divSubtractShift n d wn qr) := by
+  simp only [divSubtractShift, decide_eq_true_eq]
+  -- We add these hypotheses for `omega` to find them later.
+  have ⟨⟨hrwn, hd, hrd, hr, hn, hrnd⟩, hwn_lt⟩ := h
+  have : d.toNat * (qr.q.toNat * 2) = d.toNat * qr.q.toNat * 2 := by rw [Nat.mul_assoc]
+  by_cases rltd : shiftConcat qr.r (n.getLsb (wn - 1)) < d
+  · simp only [rltd, ↓reduceIte]
+    constructor <;> try bv_omega
+    case pos.hrWidth => apply toNat_shiftConcat_lt_of_lt_of_lt_two_pow <;> omega
+    case pos.hqWidth => apply toNat_shiftConcat_lt_of_lt_of_lt_two_pow <;> omega
+    case pos.hdiv =>
+      simp [qr.toNat_shiftRight_sub_one_eq h,
+      toNat_shiftConcat_eq_of_lt_of_lt_two_pow
+        (hk := qr.wr_lt_w h)
+        (hx := h.hrWidth),
+      h.hdiv,
+      toNat_shiftConcat_eq_of_lt_of_lt_two_pow (qr.wr_lt_w h) h.hqWidth, Nat.add_mul, toNat_false, Nat.add_zero,
+      this]
+      omega
```

Comment:
This is a surprising way to format the simp list. Can you just wrap at the end of the line.